File: day_3/second_star/rust/src/main.rs
use std::collections::VecDeque;

fn main() {
    let input = "../../input.txt";
    let input_file = std::fs::read_to_string(input).expect("Error reading input file");
    let mut engine = Engine {
        gears: Vec::new(),
        lines: VecDeque::new(),
    };
    engine.lines.push_back(("".to_string(), 0));
    for (l, line) in input_file.lines().enumerate() {
        let line = line.trim();
        if engine.lines.len() < 3 {
            engine.lines.push_back((line.to_string(), l));
            continue;
        }
        engine.parse_manual_line(line, l);
    }
    engine.gears.sort();
    engine.gears.dedup();
    println!(
        "{}",
        engine
            .gears
            .iter()
            .filter(|g| g.adjecent.len() == 2 && {
                !engine.gears.iter().any(|g2| {
                    g2.adjecent.len() > g.adjecent.len() && g2.position == g.position
                })
            })
            .map(|g| g.adjecent[0] * g.adjecent[1])
            .sum::<usize>()
    );
}
#[derive(Debug)]
struct Engine {
    gears: Vec<Gear>,
    lines: VecDeque<(String, usize)>,
}

#[derive(Debug)]
struct Number {
    value: usize,
    position: (usize, usize),
    num_of_digits: usize,
}

#[derive(Debug, PartialEq, Ord, Eq, PartialOrd)]
struct Gear {
    position: (usize, usize),
    adjecent: Vec<usize>,
}

impl Engine {
    fn parse_manual_line(&mut self, line: &str, line_num: usize) {
        self.lines.pop_front();
        self.lines.push_back((line.to_string(), line_num));
        let mut recording: bool = false;
        let mut current_nums = Vec::new();
        let mut current_gears = Vec::new();
        for (line, l) in self.lines.iter() {
            for (i, c) in line.chars().enumerate() {
                if c.is_digit(10) {
                    if !recording {
                        current_nums.push(Number {
                            value: c.to_digit(10).unwrap() as usize,
                            position: (*l, i),
                            num_of_digits: 1,
                        });
                        recording = true;
                    } else {
                        let last_num = current_nums.last_mut().unwrap();
                        last_num.value = last_num.value * 10 + c.to_digit(10).unwrap() as usize;
                        last_num.num_of_digits += 1;
                    }
                } else if c == '*' {
                    current_gears.push(Gear {
                        position: (*l, i),
                        adjecent: Vec::new(),
                    });
                    recording = false;
                } else {
                    recording = false;
                }
            }
        }
        for gear in current_gears.iter_mut() {
            for num in current_nums.iter() {
                if (gear.position.1 as i32 >= (num.position.1 as i32 - 1))
                    && ((gear.position.1 as i32) <= (num.position.1 + num.num_of_digits) as i32)
                    && ((gear.position.0 as i32) - (num.position.0 as i32)).abs() < 2
                {
                    gear.adjecent.push(num.value);
                }
            }
        }
        self.gears.extend(current_gears);
    }
}

File: day_3/second_star/zig/src/main.zig
const std = @import("std");

const Position = struct {
    x: isize,
    y: isize,
};

const Number = struct {
    value: isize,
    position: Position,
    num_of_digits: isize,
};

const Symbol = struct {
    position: Position,
};

const Gear = struct {
    position: Position,
};

const Engine = struct {
    nums: std.ArrayList(Number),
    symbols: std.ArrayList(Symbol),
    gears: std.ArrayList(Gear),

    fn parseManualLine(self: *Engine, line: []const u8, line_num: isize) !void {
        var recording: bool = false;
        for (line, 0..) |c, i| {
            if (c == '.') {
                recording = false;
                continue;
            }
            if (std.ascii.isDigit(c)) {
                if (!recording) {
                    try self.nums.append(Number{
                        .value = c - '0',
                        .position = Position{ .y = line_num, .x = @as(u32, @truncate(i)) },
                        .num_of_digits = 1,
                    });
                    recording = true;
                } else {
                    var last_num = &self.nums.items[self.nums.items.len - 1];
                    last_num.value = last_num.value * 10 + (c - '0');
                    last_num.num_of_digits += 1;
                }
            } else if (c == '*') {
                try self.gears.append(Gear{
                    .position = Position{ .y = line_num, .x = @as(u32, @truncate(i)) },
                });
                recording = false;
            } else {
                try self.symbols.append(Symbol{
                    .position = Position{ .y = line_num, .x = @as(u32, @truncate(i)) },
                });
                recording = false;
            }
        }
    }

    fn calculateGearRatios(self: *Engine) isize {
    var total: isize = 0;
    for (self.gears.items) |g| {
        var nums: [2]isize = [_]isize{0} ** 2;
        var count: isize = 0;
        for (self.nums.items) |n| {
            if ((g.position.x >= (n.position.x - 1)) and 
                (g.position.x <= (n.position.x + n.num_of_digits)) and 
                std.math.absCast(g.position.y - n.position.y) < 2) {
                nums[@as(usize, @intCast(count))] = n.value;
                count += 1;
            }
        }
        if (count == 2) {
            total += nums[0] * nums[1];
        }
    }
    return total;
}
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    var allocator = gpa.allocator();

    var engine = Engine{
        .nums = std.ArrayList(Number).init(allocator),
        .symbols = std.ArrayList(Symbol).init(allocator),
        .gears = std.ArrayList(Gear).init(allocator),
    };
    defer {
        engine.nums.deinit();
        engine.symbols.deinit();
        engine.gears.deinit();
    }

    const input_file = try std.fs.cwd().openFile("../../input.txt", .{});
    defer input_file.close();
    var reader = input_file.reader();
    var buffer: [4096]u8 = undefined;
    var line_num: isize = 0;
    while (try reader.readUntilDelimiterOrEof(&buffer, '\n')) |line| {
        try engine.parseManualLine(line, line_num);
        line_num += 1;
    }

    const output = engine.calculateGearRatios();
    std.debug.print("{}\n", .{output});
}